### PR TITLE
Registry builders

### PR DESCRIPTION
**Pull request by Machine-Maker**

### CODE REVIEW

File: patches/api/0478-Introduce-registry-entry-and-builders.patch [hunk 1]
```diff
@@ -0,0 +1,423 @@
...
++
++        /**
++         * Configures the weight of this enchantment used by the weighted random when selecting enchantments.
++         *
++         * @param weight the weight value.
++         * @return this builder.
++         */
++        @Required
++        @Contract(value = "_ -> this", mutates = "this")
++        @NonNull Builder weight(@Range(from = 1, to = 1024) int weight);
```
Comment (kennytv) [inline patches/api/0478-Introduce-registry-entry-and-builders.patch]: Either a default ot stating what a "usual" value is would be good here

File: patches/api/0478-Introduce-registry-entry-and-builders.patch [hunk 2]
```diff
@@ -0,0 +1,423 @@
...
++
++        /**
++         * Configures the maximum cost to enchant an item with this enchantment.
++         *
++         * @param maximumCost the enchantment cost.
++         * @return this builder.
++         */
++        @Required
++        @Contract(value = "_ -> this", mutates = "this")
++        @NonNull Builder maximumCost(@NotNull EnchantmentCost maximumCost);
```
Comment (kennytv) [inline patches/api/0478-Introduce-registry-entry-and-builders.patch]: Also for these two

File: patches/api/0478-Introduce-registry-entry-and-builders.patch [hunk 3]
```diff
@@ -0,0 +1,423 @@
...
++         * If the item enchanted with this enchantment is equipped in a slot not covered by the returned list and its groups,
++         * the enchantment's effects, like attribute modiifers, will not activate.
++         *
++         * @param activeSlots a list of equipment slot groups.
++         * @return this builder.
++         * @see Enchantment#getActiveSlots()
++         */
++        @Required
++        @Contract(value = "_ -> this", mutates = "this")
++        @NonNull Builder activeSlots(@NonNull List<@NonNull EquipmentSlotGroup> activeSlots);
```
Comment (kennytv) [inline patches/api/0478-Introduce-registry-entry-and-builders.patch]: Collection/Iterable?

File: patches/api/0478-Introduce-registry-entry-and-builders.patch [hunk 4]
```diff
@@ -0,0 +1,423 @@
...
++         * This enchantment may not be applied on items that are already enchanted with enchantments found in the returned
++         * set.
++         * <p>
++         * Defaults to an empty set allowing this enchantment to be applied regardless of other enchantments.
++         *
++         * @param exclusiveWith a registry set of enchantments exclusive to this one.
++         * @return this builder.
++         */
++        @Contract(value = "_ -> this", mutates = "this")
++        @NonNull Builder exclusiveWith(@NonNull RegistryKeySet<Enchantment> exclusiveWith);
```
Comment (kennytv) [inline patches/api/0478-Introduce-registry-entry-and-builders.patch]: Default to empty

File: patches/api/0478-Introduce-registry-entry-and-builders.patch [hunk 5]
```diff
@@ -0,0 +1,423 @@
...
++
++        /**
++         * Configures the set of supported items this enchantment can be applied on.
++         *
++         * @param supportedItems the tag key representing the supported items.
++         * @return this builder.
++         */
++        @Required
++        @Contract(value = "_ -> this", mutates = "this")
++        @NonNull Builder supportedItems(@NonNull RegistryKeySet<ItemType> supportedItems);
```
Comment (kennytv) [inline patches/api/0478-Introduce-registry-entry-and-builders.patch]: Linking some of of the common tag keys would be nice

**Review by lynxplay — APPROVED** [review]

**Pull request merged**